### ISSUE TITLE
Review the process by which I arrived at the two-script classifier pipeline

### ISSUE THREAD

**Issue by journeymansix** [Resilience-Biomarkers-for-Aquaculture/Cvirg_Pmarinus_RNAseq#61]
Most of the ChatGPT dialog that led to the current scripts evolved over the course of this very long chat:
https://chatgpt.com/share/68bc4d13-e418-800d-9960-f42e37d9f98b

There was lots of back and forth about formatting and types of input, and about 1/3 of the way through I had arrived at a six-gene classifier that looked very promising, and I wanted to learn how to test it on other datasets -- those from the other papers.
The gene set was not a linear model, just a set of genes with which a linear model could be trained for each of the papers. Here's the question I asked:
> So at this point (before the retraining) we think we have six genes that are good classifier *parameters*, but not a linear (or other) model that's a good classifier across studies? And we're now about to try to show that at least the *parameters* are good? Finally, what will it take to produce a unversally applicable model for C. Virginica condition?

Following this, I pursued creating a frozen model (i.e. with coefficients) and asked whether we'd need batch correction.

I then asked:
> OK, please generate code for the scheme I described: Partition each batch (A, B1, B2): Randomly hold out (say) 20–30% of samples from each batch as a final test set. Ensure the holdout is stratified by condition so both tolerant and sensitive are represented. Training phase: Pool all the remaining samples from A + B1 + B2. Apply the same preprocessing (VST, per-gene z-scores). Fit a single logistic regression using the six genes. Save the coefficients + intercept. Validation phase: Apply the frozen model to the held-out test samples (never seen in training). Report AUROC, AUPR, calibration, etc. Because the test set spans all three batches, you’re verifying both generalization across domains and predictive power on unseen individuals.

Note that up to this point I wasn't using all the collection dates from Study 5, so "the smallest batch has only 4 samples labeled with Condition=tolerant".

The next results were declared "almost too good", especially considering that test sizes were as small as 2-4 samples.
> Here are the results printed from the script. Are they good?
[SUMMARY] Overall test metrics:
{
  "n_test": 22,
  "auroc": 1.0,
  "aupr": 1.0,
  "brier": 8.116952883362103e-06,
  "log_loss": 0.0006474823838042551,
  "calibration_slope": 1.8100674581769025,
  "calibration_intercept": -0.19327904270794555,
  "auroc_ci": [
    0.9999999999999999,
    1.0
  ],
  "aupr_ci": [
    0.9999999999999999,
    1.0
  ],
  "brier_ci": [
    8.032268390722937e-11,
    2.4330239702505606e-05
  ],
  "log_loss_ci": [
    2.1488582408763657e-06,
    0.001890129077773869
  ]
}

Tried again with different seeds, and got similar results, with some of the genes very stable and others changing coefficients and sign. (We're about halfway through the chat at this point).

Note at this point in the conversation, ChatGPT had mentioned "leakage" at many points, but I hadn't followed up on it. But at this point I said:
> There are further recommendations from ChatGPT : Run stability selection inside each LOSO fold **for strict no-leakage**. Save seeds, configs, hashes for reproducibility. Evaluate calibration with reliability curves, not just AUROC. Optionally benchmark against linear SVMs. If class priors differ by batch, consider prior correction.

And so began a reworking of the scripts, starting with Stability selection inside each LOSO fold (no leakage), which caused restructuring, and doing DESeq2 within the R script rather than taking existing results. The python script that it generated, "train_eval.py" was missing some of the features of the previous python script so that required reworking to produce `lasso_prune_onefold.py`.

What follows are dozens of screenfuls of debugging. First I find the first biggest mistake ChatGPT had made:
> dude , it seems you explicity drop the gene_id column here:

... and finally
> I finally got the entire pipeline to work. How do I evaluate the results?

More follows and then I say:
> In our previous two step pipeline the second step was attached file, which did much more than train_eval.py does.

Later:
> I see that in ./loso_P&S 2023/test_metrics_overall.json, auroc is only 0.59. Should I include that fold?

After that, I bring back the idea of leakage, and say:
> The goal here is to find the minimal set of genes that can reliable be used to build a predictive classifier that can be applied to future samples to predict their resilience.

That tilts towards doing the entirely fold-separated pipeline. And this is about 3/4 through the very long chat.

After implementation back-and-forth, I arrive at:
> ./results/loso_P&S 2023/panel_candidates.txt has 8000 lines ./results/loso_P&S 2020 2015/panel_candidates.txt has 107 lines ./results/loso_P&S 2020 2017/panel_candidates.txt has 82 lines What would explain the huge difference in size of the list of candidate genes for the different folds?

It's because there's much more agreement between the two 2020 batches, than between either of those batches and 2023. Recall that they're actually measuring different traits: resistance vs. tolerance. (Note we still have the imbalance in panel_candidates length: 2000, 1120, 1219.)

And then I start exploring batch effect. I'm considering only the stated batch effect between the 2020 batches:
> If I already know that there's a strong batch effect, as reported in the paper, between the two batches of the 2020 study, what might I do differently?

This results in adding a Likelihood Ratio Test (LRT):
```
  dds_int <- DESeq(dds_full, test = "LRT",
                   reduced = ~ batch + condition, quiet = TRUE)
```

I then have it add code to compute coefficients and preds for reporting.
Current (12/05/2025) loso_intercepts_by_fold.csv:
P&S_2020_2015,5.994588118364402
P&S_2020_2017,23.339265955955195
P&S_2023,0.9961115982131414

The high positive values indicate (paraphrase): "the model predicts almost always tolerant when gene signals are neutral. Implies the tolerant class was abundant or the fold’s training data favored tolerant." 23 might be ridiculously high.
*Should revisit*

And then came my question of whether we need normalization, and an exploration based on the results of whether we need it. (There's an Issue for this.) That check is done by `4_BCa_Cl_per_fold.py`. 

I then had it do batch centering based on results, but that made the intercepts more extreme, while "predictive discrimination is unchanged" so I commented out that code.
